java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptionStateRecorder.java

```diff
@@ -0,0 +1,154 @@
+     * @return the HmacOutputStream
+     */
+    private static HmacOutputStream ensureHmacWrapsCipherStream(final OutputStream cipherStream) {
+        if (!cipherStream.getClass().equals(HmacOutputStream.class)) {
+            final String message = "Cipher lacks authentication but OutputStream is not HmacOutputStream";
+            throw new IllegalStateException(message);
+        }
+
+        return (HmacOutputStream) cipherStream;
+    }
+
+    /**
+     * Clones Cipher (and potentially HMAC) instances for future use.
+     * This should be called before data is streamed in uploadPart.
+     */
+    public EncryptionStateSnapshot record(final EncryptionState encryptionState) {
+        final HMac hmac;
+
+        if (!encryptionState.getEncryptionContext().getCipherDetails().isAEADCipher()) {
+            OutputStream cipherStream = encryptionState.getCipherStream();
+            final HmacOutputStream digestStream = ensureHmacWrapsCipherStream(cipherStream);
+            hmac = CLONER_HMAC.createClone(digestStream.getHmac());
+        } else {
+            hmac = null;
+        }
+
+        final Cipher cipher =
+                CLONER_CIPHER.createClone(encryptionState.getEncryptionContext().getCipher());
+
+        return new EncryptionStateSnapshot(encryptionState.getLastPartNumber(), cipher, hmac);
+    }
+
+    /**
+     * Restore Cipher (and potentially HMAC) instances to a provided snapshot. We want to avoid
+     * producing a new EncryptionState for a few reasons:
+     *
+     * 1. The {@code ReentrantLock} in {@code EncryptionState} is used to synchronize access
+     *    to internal encryption state, including the creation and restoration of snapshots (record and rewind).
+     * 2. Interaction between CipherOutputStream and HmacOutputStream
+     *    makes it non-trivial construct a copy of an EncryptionState that is completely separate from the original.
+     */
+    public void rewind(final EncryptionState encryptionState, final EncryptionStateSnapshot snapshot) {
+        Validate.notNull(snapshot.getCipher(),
+                "Snapshot cipher must not be null");
+        Validate.isTrue(encryptionState.getLastPartNumber() == snapshot.getLastPartNumber(),
+                "Snapshot part number must equal encryption state part number");
+        final boolean usesHmac = !encryptionState.getEncryptionContext().getCipherDetails().isAEADCipher();
+
+        final CipherOutputStream cipherStream;
+        if (usesHmac) {
+            Validate.notNull(snapshot.getHmac(), "Snapshot hmac must not be null");
+
+            final HmacOutputStream digestStream = ensureHmacWrapsCipherStream(encryptionState.getCipherStream());
+
+            try {
+                writeField(FIELD_HMACOUTPUTSTREAM_HMAC, digestStream, snapshot.getHmac());
+            } catch (IllegalAccessException e) {
+                throw new MantaReflectionException("Failed to overwrite HmacOutputStream's hmac", e);
```

Comment:
Changed `rewind` to accept the `EncryptedMultipartUpload` object instead to it can retrieve the ID in addition to the part number.